exporters/otlp/src/otlp_grpc_client.cc

```diff
@@ -140,31 +356,216 @@ OtlpGrpcClient::MakeLogsServiceStub(const OtlpGrpcClientOptions &options)
+                      nullptr);
+    }
+    return opentelemetry::sdk::common::ExportResult::kFailure;
+  }
+
+  std::shared_ptr<OtlpGrpcClientAsyncData> async_data = MutableAsyncData(options);
+  return InternalDelegateAsyncExport(async_data, stub, std::move(context), std::move(arena),
+                                     std::move(request), std::move(result_callback), logs_count,
+                                     "log(s)");
+}
+
+std::shared_ptr<OtlpGrpcClientAsyncData> OtlpGrpcClient::MutableAsyncData(
+    const OtlpGrpcClientOptions &options)
+{
+  if (!async_data_)
+  {
+    async_data_                          = std::make_shared<OtlpGrpcClientAsyncData>();
+    async_data_->export_timeout          = options.timeout;
+    async_data_->max_concurrent_requests = options.max_concurrent_requests;
+  }
+
+  return async_data_;
+}
+
+bool OtlpGrpcClient::ForceFlush(std::chrono::microseconds timeout) noexcept
+{
+  if (!async_data_)
+  {
+    return true;
+  }
+
+  std::size_t request_counter = async_data_->start_request_counter.load(std::memory_order_acquire);
+  if (request_counter <= async_data_->finished_request_counter.load(std::memory_order_acquire))
+  {
+    return true;
+  }
+
+  // ASAN will report chrono: runtime error: signed integer overflow: A + B cannot be represented
+  //   in type 'long int' here. So we reset timeout to meet signed long int limit here.
+  timeout = opentelemetry::common::DurationUtil::AdjustWaitForTimeout(
+      timeout, std::chrono::microseconds::zero());
+
+  // Wait for all the sessions to finish
+  std::unique_lock<std::mutex> lock(async_data_->session_waker_lock);
+
+  std::chrono::steady_clock::duration timeout_steady =
+      std::chrono::duration_cast<std::chrono::steady_clock::duration>(timeout);
+  if (timeout_steady <= std::chrono::steady_clock::duration::zero())
+  {
+    timeout_steady = std::chrono::steady_clock::duration::max();
+  }
+
+  while (timeout_steady > std::chrono::steady_clock::duration::zero() &&
+         request_counter > async_data_->finished_request_counter.load(std::memory_order_acquire))
+  {
+    // When changes of running_sessions_ and notify_one/notify_all happen between predicate
+    // checking and waiting, we should not wait forever.We should cleanup gc sessions here as soon
+    // as possible to call FinishSession() and cleanup resources.
+    std::chrono::steady_clock::time_point start_timepoint = std::chrono::steady_clock::now();
+    if (std::cv_status::timeout !=
+        async_data_->session_waker.wait_for(lock, async_data_->export_timeout))
+    {
+      break;
+    }
+
+    timeout_steady -= std::chrono::steady_clock::now() - start_timepoint;
+  }
+
+  return timeout_steady > std::chrono::steady_clock::duration::zero();
+}
+
+bool OtlpGrpcClient::Shutdown(std::chrono::microseconds timeout) noexcept
+{
+  if (!async_data_)
+  {
+    return true;
+  }
+
+  if (!is_shutdown_)
+  {
+    is_shutdown_ = true;
```

Comment:
Does it need thread safe?In my understand, exporters

> is this thread safe?

Good point, It's just fine when this is not thread safety but it may confuse tsan.I change it to `atomic<bool>` now.